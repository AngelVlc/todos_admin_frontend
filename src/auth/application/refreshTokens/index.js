export * from "./GetRefreshTokensUseCase";
export * from "./DeleteSelectedRefreshTokensUseCase";
