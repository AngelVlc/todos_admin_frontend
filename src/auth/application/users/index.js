export * from "./GetUserByIdUseCase";
export * from "./DeleteUserByIdUseCase";
export * from "./GetUsersUseCase";
export * from "./CreateUserUseCase";
export * from "./UpdateUserUseCase";
