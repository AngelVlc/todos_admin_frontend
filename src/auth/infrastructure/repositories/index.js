export * from './RefreshTokensRepository';
export * from './UsersRepository';
