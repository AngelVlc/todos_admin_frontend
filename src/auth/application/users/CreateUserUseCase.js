import { BaseUseCase } from "../../../shared/domain/BaseUseCase";
import { UsersRepository } from "../../infrastructure/repositories";

export class CreateUserUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new UsersRepository();

    return new CreateUserUseCase({ repository });
  }

  async execute(user) {
    return await this._repository.create(user);
  }
}
