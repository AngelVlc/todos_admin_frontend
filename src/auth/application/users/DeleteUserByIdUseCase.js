import { BaseUseCase } from "../../../shared/domain/BaseUseCase";
import { UsersRepository } from "../../infrastructure/repositories";

export class DeleteUserByIdUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new UsersRepository();

    return new DeleteUserByIdUseCase({ repository });
  }

  async execute(id) {
    return await this._repository.deleteById(id);
  }
}
