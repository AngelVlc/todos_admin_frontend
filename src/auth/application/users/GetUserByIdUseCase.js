import { BaseUseCase } from "../../../shared/domain/BaseUseCase";
import { User } from "../../domain";
import { UsersRepository } from "../../infrastructure/repositories";

export class GetUserByIdUseCase extends BaseUseCase {
  _repository;

  constructor({ repository }) {
    super();

    this._repository = repository;
  }

  static create() {
    const repository = new UsersRepository();

    return new GetUserByIdUseCase({ repository });
  }

  async execute(id) {
    const result = await this._repository.getById(id);

    return new User(result);
  }
}
