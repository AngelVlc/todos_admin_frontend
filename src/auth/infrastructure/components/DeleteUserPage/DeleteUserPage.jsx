import React, { useEffect, useState, useContext, useCallback } from "react";
import { useParams, useHistory } from "react-router-dom";
import { Breadcrumb } from "../../../../shared/infrastructure/components/Breadcrumb";
import { YesNoButtons } from "../../../../shared/infrastructure/components/YesNoButtons";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import {
  GetUserByIdUseCase,
  DeleteUserByIdUseCase,
} from "../../../application/users";

export const DeleteUserPage = () => {
  const [user, setUser] = useState();
  const { useCaseFactory } = useContext(AppContext);
  let { userId } = useParams();
  let history = useHistory();

  const getUser = useCallback(async () => {
    const getUserByIdUseCase = useCaseFactory.get(GetUserByIdUseCase);
    const user = await getUserByIdUseCase.execute(userId);

    setUser(user);
  }, [userId, useCaseFactory]);

  useEffect(() => {
    getUser();
  }, [userId, getUser]);

  const deleteUser = async () => {
    const deleteUserByIdUseCase = useCaseFactory.get(DeleteUserByIdUseCase);

    if (await deleteUserByIdUseCase.execute(userId)) {
      history.push("/users");
    }
  };

  return (
    <>
      {user && (
        <div className="container">
          <Breadcrumb
            items={[
              { url: "/users", text: "Users" },
              { url: `/users/${userId}/delete`, text: `Delete ${user.name}` },
            ]}
          />
          <h3 className="title">
            {user.isAdmin
              ? `Delete admin user '${user.name}'?`
              : `Delete user ${user.name}?`}
          </h3>
          <YesNoButtons onYes={deleteUser} onNo={() => history.push('/users')}/>
        </div>
      )}
    </>
  );
};
