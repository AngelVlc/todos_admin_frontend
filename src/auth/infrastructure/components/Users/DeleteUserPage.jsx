import React, { useEffect, useState, useContext, useCallback } from "react";
import { useParams, useHistory } from "react-router-dom";
import { Breadcrumb } from "../../../../shared/infrastructure/components/Breadcrumb/Breadcrumb";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import {
  GetUserByIdUseCase,
  DeleteUserByIdUseCase,
} from "../../../application/users";

export const DeleteUserPage = () => {
  const [user, setUser] = useState(null);
  const { useCaseFactory } = useContext(AppContext);
  let { userId } = useParams();
  let history = useHistory();

  const getUser = useCallback(async () => {
    const getUserByIdUseCase = useCaseFactory.get(GetUserByIdUseCase);
    const user = await getUserByIdUseCase.execute(userId);

    setUser(user);
  }, [userId, useCaseFactory]);

  useEffect(() => {
    getUser();
  }, [userId, getUser]);

  const deleteUser = async () => {
    const deleteUserByIdUseCase = useCaseFactory.get(DeleteUserByIdUseCase);

    if (await deleteUserByIdUseCase.execute(userId)) {
      history.push("/users");
    }
  };

  return (
    <>
      {user && (
        <div className="container">
          <Breadcrumb
            items={[
              { url: "/users", text: "Users" },
              { url: `/users/${userId}/edit`, text: user.name },
              { url: `/users/${userId}/delete`, text: `Delete ${user.name}` },
            ]}
          />
          <h3 className="title">
            {user.isAdmin
              ? `Delete admin user '${user.name}'?`
              : `Delete user ${user.name}?`}
          </h3>
          <div className="buttons">
            <button
              className="button is-danger"
              data-testid="yes"
              onClick={() => deleteUser()}
            >
              YES
            </button>
            <button
              className="button"
              data-testid="no"
              onClick={() => history.goBack()}
            >
              NO
            </button>
          </div>
        </div>
      )}
    </>
  );
};
