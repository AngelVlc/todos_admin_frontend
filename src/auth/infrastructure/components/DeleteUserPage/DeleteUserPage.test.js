import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { DeleteUserPage } from "./DeleteUserPage";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import { MemoryRouter, Route } from "react-router-dom";
import { act } from "react-dom/test-utils";
import { User } from "../../../domain";
import { GetUserByIdUseCase } from "../../../application/users";

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

const mockedDeleteUserByIdUseCase = {
  execute: jest.fn(),
};

const renderWithContextAndRouter = (isAdmin) => {
  const fakeGetUserByIdUseCase = {
    execute: () => new User({ id: 2, name: "user", isAdmin: isAdmin }),
  };

  const useCaseFactory = {
    get: (useCase) =>
      useCase == GetUserByIdUseCase
        ? fakeGetUserByIdUseCase
        : mockedDeleteUserByIdUseCase,
  };

  const context = { auth: { info: {} }, useCaseFactory };
  return {
    ...render(
      <AppContext.Provider value={context}>
        <MemoryRouter initialEntries={[`/users/2/delete`]}>
          <Route path="/users/:userId/delete">{<DeleteUserPage />}</Route>
        </MemoryRouter>
      </AppContext.Provider>
    ),
  };
};

afterEach(cleanup);

describe("DeleteUserPage", () => {
  describe("when the user is not an admin", () => {
    it("should match the snapshot when the user is not an admin", async () => {
      let fragment;
      await act(async () => {
        const { asFragment } = renderWithContextAndRouter(false);
        fragment = asFragment;
      });

      expect(fragment()).toMatchSnapshot();
    });

    it("when click on cancel should cancel the deletion", async () => {
      let container;
      await act(async () => {
        container = renderWithContextAndRouter(false);
      });

      await waitFor(() => {
        fireEvent.click(container.getByTestId("no"));
      });

      expect(mockHistoryPush).toHaveBeenCalled();
      expect(mockHistoryPush.mock.calls[0][0]).toBe("/users");
      mockHistoryPush.mockClear();
    });

    it("when click on yes should delete a user", async () => {
      let container;
      await act(async () => {
        container = renderWithContextAndRouter(false);
      });

      mockedDeleteUserByIdUseCase.execute.mockResolvedValue(true);

      await waitFor(() => {
        fireEvent.click(container.getByTestId("yes"));
      });

      expect(mockedDeleteUserByIdUseCase.execute).toHaveBeenCalled();
      expect(mockedDeleteUserByIdUseCase.execute.mock.calls[0][0]).toBe("2");
      expect(mockHistoryPush).toHaveBeenCalled();
      expect(mockHistoryPush.mock.calls[0][0]).toBe("/users");
      mockHistoryPush.mockClear();
    });
  });

  describe("when the user is an admin", () => {
    it("should match the snapshot when the user is an admin", async () => {
      let fragment;
      await act(async () => {
        const { asFragment } = renderWithContextAndRouter(true);
        fragment = asFragment;
      });

      expect(fragment()).toMatchSnapshot();
    });
  });
});
