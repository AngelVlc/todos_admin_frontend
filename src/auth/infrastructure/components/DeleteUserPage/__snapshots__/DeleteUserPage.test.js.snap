// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteUserPage when the user is an admin should match the snapshot when the user is an admin 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/users"
          >
            Users
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/users/2/delete"
          >
            Delete user
          </a>
        </li>
      </ul>
    </nav>
    <h3
      class="title"
    >
      Delete admin user 'user'?
    </h3>
    <div
      class="buttons"
    >
      <button
        class="button is-danger"
        data-testid="yes"
      >
        YES
      </button>
      <button
        class="button"
        data-testid="no"
      >
        NO
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DeleteUserPage when the user is not an admin should match the snapshot when the user is not an admin 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <nav
      aria-label="breadcrumbs"
      class="breadcrumb"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/users"
          >
            Users
          </a>
        </li>
        <li
          class="is-active"
        >
          <a
            aria-current="page"
            href="/users/2/delete"
          >
            Delete user
          </a>
        </li>
      </ul>
    </nav>
    <h3
      class="title"
    >
      Delete user user?
    </h3>
    <div
      class="buttons"
    >
      <button
        class="button is-danger"
        data-testid="yes"
      >
        YES
      </button>
      <button
        class="button"
        data-testid="no"
      >
        NO
      </button>
    </div>
  </div>
</DocumentFragment>
`;
